### PR TITLE
nixos/dae: init

### PR DESCRIPTION
**Pull request by pokon548**
## Description of changes
This pull request introduce ```services.dae``` option for existing package ```dae```. For example:
```
services.dae.enable = true;
```

As I am actively using ```dae```, I will continue to maintain this module.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/dae.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/dae.nix:3]:
```suggestion
```

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -13476,6 +13476,12 @@
     githubId = 38314551;
     name = "Peter Okelmann";
   };
+  pokon548 = {
```
Comment (AndersonTorres) [inline maintainers/maintainer-list.nix:13479]: Split.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/dae.nix (hunk 1 above)
Comment (pokon548) [reply nixos/modules/services/networking/dae.nix]: Thanks!

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (pokon548) [reply maintainers/maintainer-list.nix:13479]: Can you describe this a little bit? Did you mean I should split this change to individual pull request? Thanks

Comment (AndersonTorres) [reply maintainers/maintainer-list.nix:13479]: The maintainers list should be updated in a separate commit (no need for a PR).

File: nixos/modules/services/networking/dae.nix [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  options = {
+    services.dae = {
+      enable = lib.options.mkEnableOption (lib.mdDoc "the dae service");
+    };
+  };
+
+  config = lib.mkIf config.services.dae.enable {
+    environment.systemPackages = [ pkgs.dae ];
+
+    networking.firewall.allowedTCPPorts = [ 12345 ];
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/dae.nix:13]: This port can't be configured?

File: nixos/modules/services/networking/dae.nix [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+{
+  options = {
+    services.dae = {
+      enable = lib.options.mkEnableOption (lib.mdDoc "the dae service");
+    };
+  };
+
+  config = lib.mkIf config.services.dae.enable {
+    environment.systemPackages = [ pkgs.dae ];
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/dae.nix:11]: Use `cfg.package`.

File: nixos/modules/services/networking/dae.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        LimitNOFILE = 1048576;
+        Restart = "on-abnormal";
+        Type = "notify";
+      };
+
+      wantedBy = [ "multi-user.target" ];
+    };
+  };
+
+  meta.maintainers = with lib.maintainers; [ pokon548 ];
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/dae.nix:41]: Usually we put this as the first  attribute, in order to quickly find the culprits.

File: nixos/modules/services/networking/dae.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        After = [ "network.target" "systemd-sysctl.service" ];
+        Wants = [ "network.target" ];
+      };
+
+      serviceConfig = {
+        User = "root";
+        ExecStartPre = "${lib.getExe pkgs.dae} validate -c /etc/dae/config.dae";
+        ExecStart = "${
+            lib.getExe pkgs.dae
+          } run --disable-timestamp -c /etc/dae/config.dae";
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/dae.nix:29]:
```suggestion
        ExecStart = "${lib.getExe pkgs.dae} run --disable-timestamp -c /etc/dae/config.dae";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/dae.nix (hunk 6 above)
Comment (drupol) [reply nixos/modules/services/networking/dae.nix:29]: How about doing `ExecStart = "${lib.getExe cfp.package} run --disable-timestamp -c /etc/dae/config.dae";`

Just like it's done in [here](https://github.com/NixOS/nixpkgs/blob/aad0e18f8d86daf57cfb30e13958f70b6a73705c/nixos/modules/services/web-apps/guacamole-server.nix#L75). You will need to add the `package` config attribute, like in [here](https://github.com/NixOS/nixpkgs/blob/aad0e18f8d86daf57cfb30e13958f70b6a73705c/nixos/modules/services/web-apps/guacamole-server.nix#L13).

**Review by drupol — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/dae.nix (hunk 3 above)
Comment (pokon548) [reply nixos/modules/services/networking/dae.nix:13]: > This port can't be configured?

Yes, according to dae document: https://dae.v2raya.org/docs/current/FAQ#troubleshoot-firewall

This port is reserved for internal usage, and shall be opened by default.

File: nixos/modules/services/networking/dae.nix (hunk 6 above)
Comment (drupol) [reply nixos/modules/services/networking/dae.nix:29]: Can you please add the missing cfg.package, like it's shown in the guacamole module ? It's a one line change almost.

Comment (pokon548) [reply nixos/modules/services/networking/dae.nix]: > Can you please add the missing cfg.package, like it's shown in the guacamole module ? It's a one line change almost.

Thanks for suggestion! Already included in latest commit :)

File: nixos/modules/services/networking/dae.nix (hunk 5 above)
Comment (pokon548) [reply nixos/modules/services/networking/dae.nix:40]: Thanks!

File: nixos/modules/services/networking/dae.nix (hunk 4 above)
Comment (pokon548) [reply nixos/modules/services/networking/dae.nix]: Thanks!

**Review by drupol — CHANGES_REQUESTED** [review]
Thanks it's good here.

Last but not least, could you add `meta.mainProgram` in dae (https://github.com/NixOS/nixpkgs/blob/5faab29808a2d72f4ee0c44c8e850e4e6ada972f/pkgs/tools/networking/dae/default.nix#L57C1-L57C1) ?

Why? See this: https://github.com/NixOS/nixpkgs/pull/246386

**Comment by pokon548** [comment]
> Thanks it's good here.
> 
> Last but not least, could you add `meta.mainProgram` in dae (https://github.com/NixOS/nixpkgs/blob/5faab29808a2d72f4ee0c44c8e850e4e6ada972f/pkgs/tools/networking/dae/default.nix#L57C1-L57C1) ?
> 
> Why? See this: #246386

Thanks! Also included in latest commit :)

**Review by drupol — CHANGES_REQUESTED** [review]
All good for me.

Now the form. This PR should only contains only 3 commits:

- 1 for adding you as maintainer: `maintainers: add pokon548`
- 1 for adding the Nixos module: `nixos/dae: init`
- 1 for adding the `meta.mainProgram` in `dae`: `dae: add `meta.mainProgram`

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (pokon548) [reply maintainers/maintainer-list.nix:13479]: > The maintainers list should be updated in a separate commit (no need for a PR).

It seems already in separate commit now, thanks! :)

**Comment by pokon548** [comment]
> All good for me.
> 
> Now the form. This PR should only contains only 3 commits:
> 
> * 1 for adding you as maintainer: `maintainers: add pokon548`
> * 1 for adding the Nixos module: `nixos/dae: init`
> * 1 for adding the `meta.mainProgram` in `dae`: `dae: add `meta.mainProgram`

Thanks! Already made a force push and put all changes into 3 commits :)

**Review by drupol — APPROVED** [review]
:100: all good for me... let's wait the CI.

**Comment by AndersonTorres** [comment]
The only thing I suggest now is to invert the two latest commits, because in a certain sense the commit that sets `meta.mainProgram` is a dependency of the one creating `nixos/dae`.

**Comment by AndersonTorres** [comment]
No, it is not so necessary.

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**